Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp

```diff
@@ -110,6 +110,15 @@ InlineLayoutResult InlineFormattingContext::layoutInFlowAndFloatContent(const Co
         // FIXME: We should be able to extract the last line information and provide it to layout as "previous line" (ends in line break and inline direction).
         return PreviousLine { lastLineIndex, { }, { }, { }, { } };
     };
+
+    if (root().style().textWrap() == TextWrap::Balance) {
+        auto balanceWidth = computeTextWrapBalanceWidthConstraint(inlineItems, constraints, inlineLayoutState);
+        auto balanceHorizontalConstraints = HorizontalConstraints { constraints.horizontal().logicalLeft, constraints.horizontal().logicalWidth, balanceWidth };
+        auto balanceInFlowContentConstraints = ConstraintsForInFlowContent { balanceHorizontalConstraints, constraints.logicalTop() };
+        auto balanceConstraints = ConstraintsForInlineContent { balanceInFlowContentConstraints, constraints.visualLeft() };
+        return lineLayout(inlineItems, needsLayoutRange, previousLine(), balanceConstraints, inlineLayoutState);
+    }
```

Comment:
would just do something along the lines of

if (root().style().textWrap() == TextWrap::Balance)
  inlineLayoutState.setHorizontalConstraintOverride(computeTextWrapBalanceWidthConstraint(inlineItems, constraints, inlineLayoutState));